### PR TITLE
feat(magic-link): add handler function support for disableSignUp

### PR DESCRIPTION
**Pull request by GoPro16**
## Summary
Allow `disableSignUp` to accept a function that receives the endpoint context and returns a boolean. This enables conditional signup blocking based on request data, application state, or external conditions.

## Changes
- Updated `disableSignUp` option to accept `boolean | ((ctx: GenericEndpointContext) => boolean | Promise<boolean>)`
- Added support for both synchronous and asynchronous handler functions
- Function receives full endpoint context with access to `body`, `context`, and `request`
- Function is evaluated in both the send and verify endpoints
- Only queries the database when signup is actually disabled (efficient)

## Example Usage
```typescript
magicLink({
  sendMagicLink: async ({ email, url, token }) => {
    await sendEmail(email, url);
  },
  disableSignUp: (ctx) => {
    // Only allow signups from specific email domains
    const email = ctx.body.email;
    const allowedDomains = ["company.com", "partner.com"];
    const domain = email.split("@")[1];
    return !allowedDomains.includes(domain);
  }
})
```

Or with async logic:
```typescript
magicLink({
  disableSignUp: async (ctx) => {
    // Check feature flag or application state
    const signupsEnabled = await checkFeatureFlag("magic-link-signups");
    return !signupsEnabled;
  }
})
```

## Tests
- ✅ All existing tests pass
- ✅ Added comprehensive tests for function handler behavior
- ✅ Tests cover sync/async functions, context access, and edge cases
- ✅ Test verifies function is evaluated during verification flow
- ✅ TypeScript compilation successful with proper types

### CODE REVIEW

**Review by GoPro16 — COMMENTED** [review]
Sorry for the spam 😄

File: packages/better-auth/src/plugins/magic-link/magic-link.test.ts [hunk 1]
```diff
@@ -303,3 +303,247 @@ describe("magic link storeToken", async () => {
...
+		expect(capturedEmail).toBe(newEmail);
+	});
+
+	it("should receive request context in disableSignUp function", async () => {
+		const disableSignUpFn = vi.fn((request: Request | undefined) => {
+			// Handle undefined request case
+			if (!request) return false;
+			const referer = request.headers.get("referer");
+			return referer?.includes("blocked-path") ?? false;
+		});
```
[the remark this answers is not in the document]
Comment (GoPro16) [reply packages/better-auth/src/plugins/magic-link/magic-link.test.ts]: Thanks for the review\! Just to clarify: "referer" is actually the correct spelling for the HTTP header. It's a famous misspelling from the original HTTP specification that became standardized - the header is officially called "Referer" (not "Referrer"). See: https://en.wikipedia.org/wiki/HTTP_referer

File: docs/content/docs/plugins/magic-link.mdx [hunk 2]
```diff
@@ -134,7 +134,38 @@ and a `request` object as the second parameter.
...
+```ts
+magicLink({
+    sendMagicLink: async ({ email, token, url }, request) => {
+        // send email
+    },
+    disableSignUp: (request) => {
+        if (!request) return false;
+        const referer = request.headers.get("referer");
+        return referer?.includes("/admin") ?? false;
+    }
```
Comment (himself65) [inline docs/content/docs/plugins/magic-link.mdx]: From logical consideration, this is incorrect. You cannot trust data from http request, so disabling based on referer is insecure

Comment (GoPro16) [reply docs/content/docs/plugins/magic-link.mdx]: Agreed there, can definitely change the example. The use case we have right now is more of a UX consideration than blocking a potential attacker from spoofing the referer. ie. we want to check if the user is signing up from our sign-up page and not our sign-in page as the same `signIn.magicLink` is used for both when `disableSignUp` is `false`.

There are definitely other workarounds we can add in for now like using the `hooks` but a larger example is the `async` one where we would want to disable signup across the board but then again it also could just be in `hooks`.

File: packages/better-auth/src/plugins/magic-link/index.ts [hunk 3]
```diff
@@ -364,7 +381,17 @@ export const magicLink = (options: MagicLinkopts) => {
...
-						if (!opts.disableSignUp) {
+						// Handle disableSignUp - can be boolean or function
+						let signUpDisabled = false;
+						if (opts.disableSignUp) {
+							if (typeof opts.disableSignUp === "function") {
+								signUpDisabled = await opts.disableSignUp(ctx.request);
+							} else {
+								signUpDisabled = true;
+							}
+						}
```
Comment (Bekacru) [inline packages/better-auth/src/plugins/magic-link/index.ts]:
```suggestion
						const signUpDisabled = typeof opts.disableSignUp === "function" ? await opts.disableSignUp(ctx.request) : opts.disableSignUp
```

File: packages/better-auth/src/plugins/magic-link/index.ts [hunk 4]
```diff
@@ -29,9 +29,16 @@ interface MagicLinkopts {
...
+	 * This allows you to conditionally disable signups based on request context.
+	 *
+	 * Note: The request may be undefined in some contexts. Make sure to handle this case.
+	 *
 	 * @default false
 	 */
-	disableSignUp?: boolean;
+	disableSignUp?:
+		| boolean
+		| ((request: Request | undefined) => boolean | Promise<boolean>);
```
Comment (Bekacru) [inline packages/better-auth/src/plugins/magic-link/index.ts]: I think you should pass the `ctx` object instead. That's one of the refactors we're planning to do for other similar callback options as well

`ctx: GenericEndpointContext`

File: docs/content/docs/plugins/magic-link.mdx (hunk 2 above)
Comment (GoPro16) [reply docs/content/docs/plugins/magic-link.mdx]: Updated the example to check blocked domain list instead along side the context feedback.
https://github.com/better-auth/better-auth/pull/5089/commits/5a48c8195a4c8dc81db9ae095a29eb828ec588e9

**Comment by gustavovalverde** [comment]
@GoPro16 can you update this PR? Please ping me when you've done it :)

**Comment by GoPro16** [comment]
@gustavovalverde I am actually going to close this PR as the original intention that I had for this was correct but implemented at the wrong level. Reason being that if we are trying to "block" domains typically we would not want to expose as a raw error of "signup is disabled" or "email is blocked" because of an attacker gaining insight easily that the policy exists — and from there enumerating which domains/emails the policy applies to. We also pivoted to using email-otp and handling similar logic in a before hook.
